File: rust/src/timer.rs
use crate::term::{TermString, ToTermString};
use std::cell::RefCell;
use std::collections::VecDeque;
use std::time::{Duration, Instant};

struct LimitedDeque<T> {
    data: VecDeque<T>,
    max_len: usize,
}

impl<T> LimitedDeque<T> {
    fn new(max_len: usize) -> Self {
        LimitedDeque {
            data: VecDeque::new(),
            max_len,
        }
    }

    fn back(&self) -> Option<&T> {
        self.data.back()
    }

    fn clear(&mut self) {
        self.data.clear()
    }

    fn push_back(&mut self, value: T) -> Option<T> {
        self.data.push_back(value);
        if self.data.len() > self.max_len {
            self.data.pop_front()
        } else {
            None
        }
    }
}

pub struct Timer {
    timezero: Instant,
    records: RefCell<LimitedDeque<u128>>,
}

impl Timer {
    pub const NANOS_PER_SEC: u128 = Duration::from_secs(1).as_nanos();
    pub const NANOS_PER_MILLI: u128 = Duration::from_millis(1).as_nanos();
    pub const NANOS_PER_MICRO: u128 = Duration::from_micros(1).as_nanos();

    const FMT_SEP: &str = " - ";

    pub fn new() -> Self {
        Self {
            timezero: Instant::now(),
            records: RefCell::new(LimitedDeque::new(100_000)),
        }
    }

    pub fn records(&self) -> VecDeque<u128> {
        self.records.borrow().data.clone()
    }

    pub fn check(&self, record: bool) -> u128 {
        let elapsed = self.timezero.elapsed().as_nanos();
        if record {
            self.records.borrow_mut().push_back(elapsed);
        }
        elapsed
    }

    pub fn check_fmt(&self, record: bool) -> TermString {
        let mut nanos = self.check(record);

        let secs = nanos / Self::NANOS_PER_SEC;
        nanos %= Self::NANOS_PER_SEC;

        let millis = nanos / Self::NANOS_PER_MILLI;
        nanos %= Self::NANOS_PER_MILLI;

        let micros = nanos / Self::NANOS_PER_MICRO;
        nanos %= Self::NANOS_PER_MICRO;

        secs.to_tstr().rjust(3)
            + "s".to_tstr().set_dim()
            + Self::FMT_SEP
            + millis.to_tstr().rjust(3)
            + "ms".to_tstr().set_dim()
            + Self::FMT_SEP
            + micros.to_tstr().rjust(3)
            + "μs".to_tstr().set_dim()
            + Self::FMT_SEP
            + nanos.to_tstr().rjust(3)
            + "ns".to_tstr().set_dim()
    }

    pub fn check_delta(&self, record: bool) -> u128 {
        let elapsed = self.check(record);
        if let Some(&latest) = self.records.borrow_mut().back() {
            elapsed - latest
        } else {
            elapsed
        }
    }

    pub fn reset(&mut self) -> &mut Self {
        self.timezero = Instant::now();
        self.records.borrow_mut().clear();
        self
    }
}

File: python/src/lifegame/timer.py
import time
from collections import deque
from typing import Self

from .term import TermString


class Timer:
    NANOS_PER_SEC = 10**9
    NANOS_PER_MILLI = 10**6
    NANOS_PER_MICRO = 10**3

    FMT_SEP = " - "

    def __init__(self, records_capacity: int = 100_000) -> None:
        self.__timezero = time.time_ns()
        self.__records: deque[int] = deque(maxlen=records_capacity)

    @property
    def records(self) -> deque[int]:
        return self.__records

    def check(self, record: bool = False) -> int:
        elapsed = time.time_ns() - self.__timezero
        if record:
            self.__records.append(elapsed)
        return elapsed

    @staticmethod
    def _measurement_fmt(value: int, unit: str) -> TermString:
        return f"{value:>3} " + TermString(unit).set_dim()

    @classmethod
    def format(cls, nanos: int) -> TermString:
        secs, nanos = divmod(nanos, cls.NANOS_PER_SEC)
        millis, nanos = divmod(nanos, cls.NANOS_PER_MILLI)
        micros, nanos = divmod(nanos, cls.NANOS_PER_MICRO)
        return (
            cls._measurement_fmt(secs, "s")
            + cls.FMT_SEP
            + cls._measurement_fmt(millis, "ms")
            + cls.FMT_SEP
            + cls._measurement_fmt(micros, "μs")
            + cls.FMT_SEP
            + cls._measurement_fmt(nanos, "ns")
        )

    def check_fmt(self, record: bool = False) -> TermString:
        return self.format(self.check(record))

    def check_delta(self, record: bool = False) -> int:
        elapsed = self.check(record)
        if self.__records:
            delta = elapsed - self.__records[-1]
        else:
            delta = elapsed
        return delta

    def reset(self) -> Self:
        self.__timezero = time.time_ns()
        self.__records.clear()
        return self
